assimilation_code/modules/assimilation/assim_tools_mod.f90

```diff
@@ -879,355 +690,74 @@ subroutine filter_assim(ens_handle, obs_ens_handle, obs_seq, keys,           &
-
-      ! Only have to update obs that have not yet been used
-      if(my_obs_indx(obs_index) > i) then
-
-         ! If the forward observation operator failed, no need to
-         ! update the unassimilated observations
-         if (any(obs_ens_handle%copies(1:ens_size, obs_index) == MISSING_R8)) cycle OBS_UPDATE
-
-         ! Compute the distance and the covar_factor
-         cov_factor = comp_cov_factor(close_obs_dist(j), cutoff_rev, &
-            base_obs_loc, base_obs_type, my_obs_loc(obs_index), my_obs_kind(obs_index))
-
-         ! if external impact factors supplied, factor them in here
-         ! FIXME: this would execute faster for 0.0 impact factors if
-         ! we check for that before calling comp_cov_factor.  but it makes
-         ! the logic more complicated - this is simpler if we do it after.
-         if (adjust_obs_impact) then
-            impact_factor = obs_impact_table(base_obs_type, my_obs_kind(obs_index))
-            cov_factor = cov_factor * impact_factor
-         endif
-
-         if(cov_factor <= 0.0_r8) cycle OBS_UPDATE
 
-         if (timing(SM_GRN)) call start_timer(t_base(SM_GRN), t_items(SM_GRN), t_limit(SM_GRN), do_sync=.false.)
-         ! Loop through and update ensemble members in each group
-         do group = 1, num_groups
-            grp_bot = grp_beg(group)
-            grp_top = grp_end(group)
-            call update_from_obs_inc(obs_prior(grp_bot:grp_top), obs_prior_mean(group), &
-               obs_prior_var(group), obs_inc(grp_bot:grp_top), &
-                obs_ens_handle%copies(grp_bot:grp_top, obs_index), grp_size, &
-                increment(grp_bot:grp_top), reg_coef(group), net_a(group))
-         end do
-         if (timing(SM_GRN)) call read_timer(t_base(SM_GRN), 'update_from_obs_inc_O', &
-                                             t_items(SM_GRN), t_limit(SM_GRN), do_sync=.false.)
+   if(.not. inflate_only) then
+      ! Now everybody updates their obs priors (only ones after this one)
+      OBS_UPDATE: do j = 1, num_close_obs
+         obs_index = close_obs_ind(j)
 
-         ! FIXME: could we move the if test for inflate only to here?
+         ! Only have to update obs that have not yet been used
+         if(my_obs_indx(obs_index) > i) then
 
-         ! Compute an information factor for impact of this observation on this state
-         if(num_groups == 1) then
-             reg_factor = 1.0_r8
-         else
-            ! Pass the time along with the index for possible diagnostic output
-            ! Compute regression factor for this obs-state pair
-            ! Negative indicates that this is an observation index
-            reg_factor = comp_reg_factor(num_groups, reg_coef, obs_time, i, -1*my_obs_indx(obs_index))
-         endif
+            ! If forward observation operator failed, no need to update unassimilated observations
+            if (any(obs_ens_handle%copies(1:ens_size, obs_index) == MISSING_R8)) cycle OBS_UPDATE
```

Comment:
The obs_qc is stored by the process that owns an observation and is broadcast when that observation is assimilated. The only observations being updated are the ones that have not yet had the qc broadcast, so that information is not available here.